test/extensions/common/wasm/wasm_test.cc

```diff
@@ -939,6 +953,275 @@ TEST_P(WasmCommonTest, RemoteCodeMultipleRetry) {
   dispatcher->clearDeferredDeleteList();
 }
 
+// test that wasm imports/exports do not work when ABI restriction is enforced
+TEST_P(WasmCommonTest, RestrictCapabilities) {
+  if (GetParam() == "null") {
+    return;
+  }
+  Stats::IsolatedStoreImpl stats_store;
+  Api::ApiPtr api = Api::createApiForTest(stats_store);
+  Upstream::MockClusterManager cluster_manager;
+  Event::DispatcherPtr dispatcher(api->allocateDispatcher("wasm_test"));
+  auto scope = Stats::ScopeSharedPtr(stats_store.createScope("wasm."));
+  NiceMock<LocalInfo::MockLocalInfo> local_info;
+  auto name = "";
+  auto root_id = "";
+  auto vm_id = "";
+  auto vm_configuration = "restrict_all";
+  auto plugin_configuration = "";
+  const auto code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/common/wasm/test_data/test_restriction_cpp.wasm"));
+  EXPECT_FALSE(code.empty());
+  auto plugin = std::make_shared<Extensions::Common::Wasm::Plugin>(
+      name, root_id, vm_id, GetParam(), plugin_configuration, false,
+      envoy::config::core::v3::TrafficDirection::UNSPECIFIED, local_info, nullptr);
+  auto vm_key = proxy_wasm::makeVmKey(vm_id, vm_configuration, code);
+
+  // restriction enforced if allowed_capabilities is non-empty
+  proxy_wasm::AllowedCapabilitiesMap allowed_capabilities{
+      {"foo", proxy_wasm::SanitizationConfig()}};
+  auto wasm = std::make_unique<Extensions::Common::Wasm::Wasm>(
+      absl::StrCat("envoy.wasm.runtime.", GetParam()), vm_id, vm_configuration, vm_key,
+      allowed_capabilities, scope, cluster_manager, *dispatcher);
+
+  EXPECT_FALSE(wasm->capabilityAllowed("proxy_on_vm_start"));
+  EXPECT_FALSE(wasm->capabilityAllowed("proxy_log"));
+  EXPECT_FALSE(wasm->capabilityAllowed("fd_write"));
+
+  EXPECT_NE(wasm, nullptr);
+  auto context = std::make_unique<TestContext>(wasm.get());
+  EXPECT_TRUE(wasm->initialize(code, false));
+
+  // on_vm_start will trigger proxy_log, but expect no call because both are restricted
```

Comment:
Should `proxy_on_vm_start` callback be even called considering that is not allowed? This is special case, since if we don't call `proxy_on_vm_start` then there is a high chance that plugin won't work correctly, but other callbacks should be skipped if they are not in the allowed capabilities.

Reply:
Yeah, the comment is incorrect, proxy_on_vm_start will never run in this case. Fixed. I also added a different callback to the wasm file (proxy_on_context_create) and adjusted the tests to test for it.